full-service/src/service/t3_sync.rs

```diff
@@ -0,0 +1,152 @@
+use mc_common::logger::{log, Logger};
+
+use t3_api::{T3Uri, TransparentTransaction};
+use t3_connection::T3Connection;
+
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+};
+
+use clap::Parser;
+
+use crate::{
+    db::{
+        models::{AuthenticatedSenderMemo, Txo},
+        txo::{TxoMemo, TxoModel},
+        Conn,
+    },
+    error::T3SyncError,
+    WalletDb,
+};
+
+#[derive(Clone, Debug, Parser)]
+pub struct T3Config {
+    #[clap(long, env = "T3_URI")]
+    pub t3_uri: T3Uri,
+    #[clap(long, env = "T3_API_KEY")]
+    pub t3_api_key: String,
+}
+
+// How many txos to sync per round
+const TXO_CHUNK_SIZE: usize = 5;
+// How long to wait in milliseconds between sync rounds
+const T3_SYNC_INTERVAL: u64 = 1000;
+
+/// T3 Sync thread - holds objects needed to cleanly terminate the t3 sync
+/// thread.
+pub struct T3SyncThread {
+    /// The main sync thread handle.
+    join_handle: Option<thread::JoinHandle<()>>,
+
+    /// Stop trigger, used to signal the thread to terminate.
+    stop_requested: Arc<AtomicBool>,
+}
+
+impl T3SyncThread {
+    pub fn start(config: T3Config, wallet_db: WalletDb, logger: Logger) -> Self {
+        let stop_requested = Arc::new(AtomicBool::new(false));
+        let thread_stop_requested = stop_requested.clone();
+
+        let t3_connection =
+            T3Connection::new(&config.t3_uri, config.t3_api_key.clone(), logger.clone());
+
+        let join_handle = Some(
+            thread::Builder::new()
+                .name("t3_sync".to_string())
+                .spawn(move || {
+                    log::debug!(logger, "T3 Sync thread started.");
+
+                    let conn = &mut wallet_db
+                        .get_pooled_conn()
+                        .expect("failed getting wallet db connection");
+
+                    loop {
+                        if thread_stop_requested.load(Ordering::SeqCst) {
+                            log::debug!(logger, "T3SyncThread stop requested.");
+                            break;
+                        }
+
+                        match sync_txos(conn, &t3_connection, &logger) {
+                            Ok(()) => (),
+                            Err(e) => log::error!(&logger, "Error during t3 sync:\n{:?}", e),
+                        }
+
+                        // This sleep is to allow other API calls that need access to the database a
+                        // chance to execute, because the t3 sync process requires a write lock on
```

Comment:
where is this write lock?

Reply:
So, we had an issue during the account sync where if we didn't add a slight delay to the syncing, it would effectively keep relocking the database until the entire account was fully synced without ever letting anything else access the DB. I'm guessing the same thing is going to happen here, even though it's only locked during the single update (unlike the sync thread which creates an exclusive transaction for a large chunk of its operation).

I think 1000ms is actually too long and have lowered it to 10ms similar to the sync function, which should alleviate any potential issues.

Let me know if you have any thoughts on this approach though?

Reply:
I think the only reason the sleep is needed here is because we don't know if we have more transactions to sync or not and without sleep this loop would just take 100% CPU.